Hint Generation Re-work and Distribution Plando Support

mracsys opened the pull request:
This is a rollup of several hint-related changes:

* Hint distributions defined in external files (`data_path/Hints`)
* GUI hint distribution list dynamically generated from available distributions
* Hint group inclusion and exclusion by location or filled item
* Required item location hints for bingo via plando
* Dungeon hint limits defined with distributions instead of hard-coded in the functions
* Custom hint text for always and sometimes hint groups, including sometimes sub-types
* Plando support for all of the above

New plando settings to control hint generation are `item_hints` and `hint_dist_user`.

## Hint Distribution Files ##
Hint distribution file format is similar to the old hard-coded dicts, plus new keys to offer finer control of different hint groups. The goal of this change is to facilitate easier sharing and updating of custom hint distributions for the different racing tournaments. An example with the Tournament distribution is provided below.

```
{
    "name":                  "tournament",
    "gui_name":              "Tournament",
    "description":           "Fixed number of hints for each type, contains duplicates, and only useful hints.",
    "add_locations":         [
        { "location": "Deku Theater Skull Mask", "types": ["always"] }
    ],
    "remove_locations":      [],
    "add_items":             [],
    "remove_items":          [],
    "dungeons_woth_limit":   2,
    "dungeons_barren_limit": 1,
    "named_items_required":  "True",
    "distribution":          {
        "trial":      {"order":  1, "weight": 0.0, "fixed":   0, "stones": 2},
        "always":     {"order":  2, "weight": 0.0, "fixed":   0, "stones": 2},
        "woth":       {"order":  3, "weight": 0.0, "fixed":   5, "stones": 2},
        "barren":     {"order":  4, "weight": 0.0, "fixed":   3, "stones": 2},
        "entrance":   {"order":  5, "weight": 0.0, "fixed":   4, "stones": 2},
        "sometimes":  {"order":  6, "weight": 0.0, "fixed": 100, "stones": 2},
        "random":     {"order":  7, "weight": 9.0, "fixed":   0, "stones": 2},
        "item":       {"order":  0, "weight": 0.0, "fixed":   0, "stones": 2},
        "song":       {"order":  0, "weight": 0.0, "fixed":   0, "stones": 2},
        "overworld":  {"order":  0, "weight": 0.0, "fixed":   0, "stones": 2},
        "dungeon":    {"order":  0, "weight": 0.0, "fixed":   0, "stones": 2},
        "junk":       {"order":  0, "weight": 0.0, "fixed":   0, "stones": 2},
        "named-item": {"order":  0, "weight": 0.0, "fixed":   0, "stones": 2}
    }
}
```

GUI names and tooltip descriptions are now defined with each distribution file. Tooltips are automatically parsed to add line returns as needed for the GUI.

The `distribution` key is a similar format to the previous dicts. Each hint group must be listed with the following keys:

* `order`: Integer to sort hint types from smallest to largest
* `weight`: Same as hint type weight in balanced and similar hint distributions
* `fixed`: Same as tournament distribution number of hints per world
* `stones`: Same as count per hint in predefined distributions

Tournament-style fixed hints and weighted-style hints are defined together in the same distribution. Using the tournament distribution as an example, fixed hints are rolled according to the defined order and fixed number. If any gossip stones remain, the `random` hint group is rolled as part of the weighted hint types (in this case the only one, so it is guaranteed upon exhaustion of all sometimes hints). Other built-in hint types with no fixed hints have non-zero orders for all types and weights equal to the old style weight, and all types have a `fixed` value of `0`.

Dungeon way of the hero and barren limitations are moved out of the hint generation functions into the distribution files. Values are identical in all built-in distributions except `very_strong`, which has a limit of `40` for both (equivalent to no limit).

Four new lists are provided to include or exclude hinted locations from different hint groups, as specified by the location name or filled item name. As seen in the tournament distribution, the `add_locations` key can be used to force a given location to be hinted by adding it to the `always` hint type. Locations can be added to (or removed from) any hint type. Likewise, the item keys allow for adding or removing locations to be hinted based on what items they contain. For example, to remove all songs from being hinted as way of the hero, the following would be utilized:

```
"remove_items": [
    { "item": "Zeldas Lullaby",     "types": ["woth"] },
    { "item": "Eponas Song",        "types": ["woth"] },
    { "item": "Sarias Song",        "types": ["woth"] },
    { "item": "Suns Song",          "types": ["woth"] },
    { "item": "Song of Time",       "types": ["woth"] },
    { "item": "Song of Storms",     "types": ["woth"] },
    { "item": "Minuet of Forest",   "types": ["woth"] },
    { "item": "Bolero of Fire",     "types": ["woth"] },
    { "item": "Serenade of Water",  "types": ["woth"] },
    { "item": "Requiem of Spirit",  "types": ["woth"] },
    { "item": "Nocturne of Shadow", "types": ["woth"] },
    { "item": "Prelude of Light",   "types": ["woth"] }
]
```

Alternately, if songsanity is disabled, the song locations could be removed from the `woth` hint type to achieve the same effect.

Barren hints are the only hint type that cannot have locations added to or removed via `add_locations` or `remove_locations`. This is to prevent situations where required items are in hinted barren areas. Using the song example again, a foolish Song of Storms that leads to a required item in Bottom of the Well would be confusing to players unfamiliar with the setting. Excluding from way of the hero means it won't be hinted, not that the game is actively lying to you. However, caution should be exercised in adding items as hinted way of the hero outside of meme settings.

Barren hints can be added or removed from consideration for specific items via the `add_items` and `remove_items` lists. This affects not just hints, but also the `:barren_regions` list in the spoiler log. In contrast, way of the hero items will still be listed in the `:woth_locations` section regardless of hint distribution changes. Forced barren items that are logically required still show up in the `:woth_locations` list if required by the seed.

## item_hints ##
Contains a list of items as named in ItemList.py to be added as "random" location hints. Built-in hint distributions assume stones/hint is the same as the sometimes category equivalent of the original distribution. The `named_items_required` hint distribution key controls if named item hints are treated as always hints. In contrast to always hints, an error will be thrown if any required named item hint cannot be placed.

Plando or third party hint distribution files can specify fixed or weighted `named-item` group hints if `named_items_required` is false.

Note that `add_items` provides similar functionality to the `item_hints` plando key if an item is added to the `always` group. The reason for duplication of this feature is to provide an easier interface for Bingo races desiring hinted tile goals without building a whole hint distribution from scratch.

## hint_dist_user ##
This setting takes a hint distribution dictionary as defined by the built-in hint distribution files. If specified, it overrides the GUI-selected hint distribution. Internally, the distribution is defined as `custom`.

## Hint Generation ##
Fixed hints are rolled first. Any remaining gossip stones are filled from a random choice of hint types with assigned weights. Checks exist to make sure there are enough gossip stones left for a given random hint type count per hint. Fixed hints do not have this safeguard and rely on the user not to overfill them.

## Example Plando ##
The attached plando

* Requires a "random"-type location hint for the location containing Lens of Truth
* Adds shooting the sun in Lake Hylia as an always hint with custom text
* Removes Frog Ocarina Game and 40 Skulls as always hints
* Keeps other "conditional" always hints from Standard weekly settings, including Skull Mask
* Adds the location with Zelda's Lullaby as an always hint
* Removes all other song locations from way of the hero and sometimes hint types
* Forces Rutos Letter to be considered foolish, and places it outside Ganon's Castle to verify
* Excludes Biggoron Sword from foolish hints, and places it in the Haunted Wasteland to verify

and applies a custom hint distribution as follows:

#### Fixed Hints ####
(in order)
* Always and Trials hints at 2 stones/hint
* 1 Named Item hint for Lens of Truth at 2 stones/hint (as required by `named_items_required`)
* 3 Way of the Hero hints at 2 stones/hint
* 2 Foolish hints at 1 stone/hint
* 4 Entrance hints at 2 stones/hint
* 5 Sometimes hints at 1 stone/hint

#### Weighted Hints ####
On weekly settings, this is 27-29 gossip stones filled, depending on if Zelda's Lullaby is on Ocarina of Time. The remaining 11-13 stones draw hints from weighted hint types as follows:

* Random (0.5 weight, 2 stones/hint)
* Item (1.0 weight, 2 stones/hint)
* Song (2.0 weight, 2 stones/hint)
* Overworld (3.0 weight, 2 stones/hint)
* Dungeon (1.0 weight, 2 stones/hint)
* Junk (0.5 weight, 1 stone/hint)

At least 1 junk hint will be added due to the odd number of remaining stones.
```
{
  "settings":            {
    "item_hints": [
      "Lens of Truth"
    ],
    "hint_dist_user": {
      "name":                  "custom",
      "gui_name":              "Plando Test Hint Distribution",
      "description":           "for debugging",
      "add_locations":         [
          { "location": "LH Sun",                  "types": ["always"], "text": "those who #blot out the sun# will find" },
          { "location": "Deku Theater Skull Mask", "types": ["always"] }
      ],
      "remove_locations":      [
          { "location": "ZR Frogs Ocarina Game",        "types": ["always"] },
          { "location": "Kak 40 Gold Skulltula Reward", "types": ["always"] }
      ],
      "add_items":             [
          { "item": "Zeldas Lullaby",     "types": ["always"] },
          { "item": "Rutos Letter",       "types": ["barren"] }
      ],
      "remove_items":          [
          { "item": "Eponas Song",        "types": ["woth", "sometimes"] },
          { "item": "Sarias Song",        "types": ["woth", "sometimes"] },
          { "item": "Suns Song",          "types": ["woth", "sometimes"] },
          { "item": "Song of Time",       "types": ["woth", "sometimes"] },
          { "item": "Song of Storms",     "types": ["woth", "sometimes"] },
          { "item": "Minuet of Forest",   "types": ["woth", "sometimes"] },
          { "item": "Bolero of Fire",     "types": ["woth", "sometimes"] },
          { "item": "Serenade of Water",  "types": ["woth", "sometimes"] },
          { "item": "Requiem of Spirit",  "types": ["woth", "sometimes"] },
          { "item": "Nocturne of Shadow", "types": ["woth", "sometimes"] },
          { "item": "Prelude of Light",   "types": ["woth", "sometimes"] },
          { "item": "Biggoron Sword",     "types": ["barren"] }
      ],
      "dungeons_woth_limit":   2,
      "dungeons_barren_limit": 1,
      "named_items_required":  "True",
      "distribution":          {
          "trial":      {"order":  1, "weight": 0.0, "fixed":   0, "stones": 2},
          "always":     {"order":  2, "weight": 0.0, "fixed":   0, "stones": 2},
          "woth":       {"order":  3, "weight": 0.0, "fixed":   3, "stones": 2},
          "barren":     {"order":  4, "weight": 0.0, "fixed":   2, "stones": 1},
          "entrance":   {"order":  5, "weight": 0.0, "fixed":   4, "stones": 2},
          "sometimes":  {"order":  6, "weight": 0.0, "fixed":   5, "stones": 1},
          "random":     {"order":  7, "weight": 0.5, "fixed":   0, "stones": 2},
          "item":       {"order":  8, "weight": 1.0, "fixed":   0, "stones": 2},
          "song":       {"order":  9, "weight": 2.0, "fixed":   0, "stones": 2},
          "overworld":  {"order": 10, "weight": 3.0, "fixed":   0, "stones": 2},
          "dungeon":    {"order": 11, "weight": 1.0, "fixed":   0, "stones": 2},
          "junk":       {"order": 12, "weight": 0.5, "fixed":   0, "stones": 1},
          "named-item": {"order": 13, "weight": 0.0, "fixed":   0, "stones": 2}
      }
    }
  },
  "randomized_settings": {},
  "starting_items":      {},
  "item_pool":           {},
  "entrances":           {},
  "locations":           {
      "OGC Great Fairy Reward": "Rutos Letter",
      "Wasteland Chest": "Biggoron Sword"
  }
}
```

r0bd0g commented:
I'm curious what happens if you try to set an always hint for an unhinted location? I wonder if every location ought to have a hint text so that you could specify whatever you want there and still get hints for it.

mracsys commented:
Custom always hints for any unhinted location silently fail due to how getHintGroup works. If it's not in the hintTable, it's hint type isn't checked.

It would be neat to have all of the locations in there with null hint groups to get more control. What I did for always hints could be applied to other hints types, and `exclude_xxxx_hints` settings could be added to remove locations from a type pool without having to list all of them. However, I'm not up for writing all of the in-game text.

DobbyDigital commented:
Is it possible to supply a list of items to be always hint? So instead of saying "tell me what's at shooting the sun" you tell it "tell me where Megaton Hammer is"? I did a quick look through the hint functions, and I don't see one that tells you where a specific item is located. If that were to exist, it could easily feed into the always_hints_user list quite easily. Not sure if you've encountered a function like that, or can think of a different way of doing it.

mracsys commented:
Not currently. Reverse lookup by item for hint plando is on my radar though. This would need some sort of count input for progressive items. For example, a count of 2 for progressive strength upgrade would be equivalent to hinting silver gauntlets.

There would not be an easy way to guarantee the hinted item is the intended progression. Going back to the silvers example, the game might hint one upgrade in Kakariko and one locked in Ganon's Castle behind Spirit-locked bridge conditions. I see this as more of a feature personally.

DobbyDigital commented:
I was viewing it more from the point of view of bingo, where "progression" is just a suggestion anyway Kappa The counts for progressive items would also be nice for board-specific goals, like if "Bullet Bag 40" were to be on the board

mracsys commented:
Reverting to draft for now as I'm looking to add a few more options. I'll re-open when those are done.

mracsys commented:
While developing this I noticed a couple location names in `HintList.py` did not match `LocationList.py`. Both had escaped apostrophes in the name in `HintList.py`. I fixed it in the current draft PR.

I am also taking the opportunity to add all remaining locations to `hintTable` to improve usability of the new add/remove keys. I'll get to this in the coming week.

r0bd0g commented:
What was the practical effect of those locations names not matching?

mracsys commented:
Yes, they get excluded by `hintExclusions`, and as far as I can tell this isn't logged. Probably wasn't caught because the locations were the Fountain adult skull and an MQ DC skull.

mracsys commented:
Entrance hints are currently broken. #1078 lays the groundwork to fix the key conflicts in the hint table. If/when that is merged, I'll update my hint table to match the new location and region names as well as restore the region hint strings.

mracsys commented:
New location names are merged in and entrance hints are tested. This should be ready for review now (for real, sorry about that!)

Hamsda commented:
Gotta say, I love the idea of this. Great work!

Zannick commented:
Could you change the "stones" parameter to "copies"? Would be a bit less ambiguous (not that anyone would confuse this with spiritual stones >.>)

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Jabu Jabus Belly MQ Second Room Upper Chest':                 ("#moving anatomy# creates a path to", None, 'exclude'),
+    'Jabu Jabus Belly MQ Basement Near Switches Chest':            ("a #pair of digested cows# hold", None, 'exclude'),
+    'Jabu Jabus Belly MQ Basement Near Vines Chest':               ("a #pair of digested cows# hold", None, 'exclude'),
+    'Jabu Jabus Belly MQ Near Boss Chest':                         ("the #final cows' reward# is", None, 'exclude'),
```
I think this hint ought to be clearer it's about Jabu Jabu (and not say, 5 cows on ZR wanting to hear songs).

mracsys replied:
Done

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Jabu Jabus Belly MQ Basement Near Switches Chest':            ("a #pair of digested cows# hold", None, 'exclude'),
+    'Jabu Jabus Belly MQ Basement Near Vines Chest':               ("a #pair of digested cows# hold", None, 'exclude'),
+    'Jabu Jabus Belly MQ Near Boss Chest':                         ("the #final cows' reward# is", None, 'exclude'),
+    'Jabu Jabus Belly MQ Falling Like Like Room Chest':            ("#trapped cows# guard", None, 'exclude'),
```
Same here. (Aren't most cows trapped?)

mracsys replied:
Done

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Water Temple MQ Map Chest':                                   ("#sparring soldiers# in the Water Temple guard", None, 'exclude'),
+
+    'Spirit Temple Child Bridge Chest':                            ("a child conquers a #skull in green fire# in the Spirit Temple to reach", None, 'exclude'),
+    'Spirit Temple Child Early Torches Chest':                     ("a child can find a #caged chest in Spirit Temple# with", None, 'exclude'),
```
"in the Spirit Temple"?

mracsys replied:
moved text outside of the hashes to be consistent with other hints

Zannick replied:
I meant there should be a "the" to be consistent with other hints.

mracsys replied:
Whoops! My brain skipped right over that. Fixed now.

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Spirit Temple MQ Child Climb South Chest':                    ("#trapped by falling enemies# within the Spirit Temple is", None, 'exclude'),
+    'Spirit Temple MQ Compass Chest':                              ("#blinding the colossus# unveils", None, 'exclude'),
+    'Spirit Temple MQ Statue Room Lullaby Chest':                  ("a #royal melody awakens the colossus# to reveal", None, 'exclude'),
+    'Spirit Temple MQ Statue Room Invisible Chest':                ("the #Eye of Truth# finds the colossus' hidden", None, 'exclude'),
```
"colossus's"

mracsys replied:
Done

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Shadow Temple MQ Invisible Spikes Chest':                     ("the #dead roam among invisible spikes# guarding", None, 'exclude'),
+    'Shadow Temple MQ Boss Key Chest':                             ("#walls consumed by a ball of fire# reveal", None, 'exclude'),
+    'Shadow Temple MQ Spike Walls Left Chest':                     ("#walls consumed by a ball of fire# reveal", None, 'exclude'),
+    'Shadow Temple MQ Stalfos Room Chest':                         ("near an #empty pedastal# within the Shadow Temple lies", None, 'exclude'),
```
pedestal

mracsys replied:
Done

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Ice Cavern Iron Boots Chest':                                 ("a #monster in a frozen cavern# guards", None, 'exclude'),
+    'Ice Cavern Freestanding PoH':                                 ("a #wall of ice# protects", None, 'exclude'),
+
+    'Ice Cavern MQ Iron Boots Chest':                              ("a #monster in a frozen cavern# guard", None, 'exclude'),
```
guards

mracsys replied:
Done

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+    'Fire Temple Volvagia Heart':                                  ("the #Subterranean Lava Dragon# guards", "#Volvagia# guards", 'exclude'),
+    'Water Temple Morpha Heart':                                   ("the #Giant Aquatic Amoeba# guards", "#Morpha# guards", 'exclude'),
+    'Spirit Temple Twinrova Heart':                                ("the #Sorceress Sisters# hold", "#Twinrova# guards", 'exclude'),
+    'Shadow Temple Bongo Bongo Heart':                             ("the #Phantom Shadow Beast# guards", "#Bongo Bongo# guards", 'exclude'),
```
Should these all be hold instead of guard?

mracsys replied:
Changed to "holds" for all. I can go either way on these.

Zannick commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+
+    'Ice Cavern MQ GS Scarecrow':                                  ("a #spider above icy pits# holds", None, 'exclude'),
+    'Ice Cavern MQ GS Ice Block':                                  ("a #web of ice# surrounds a spider with", None, 'exclude'),
+    'Ice Cavern MQ GS Red Ice':                                    ("a #spider in firey ice# hoards", None, 'exclude'),
```
fiery. or crimson.

mracsys replied:
Changed to fiery

Zannick commented on Hints.py:
```diff
@@ -284,7 +284,10 @@ def get_woth_hint(spoiler, world, checked):
-        not (world.woth_dungeon >= 2 and location.parent_region.dungeon), 
+        not (world.woth_dungeon >= world.hint_dist_user['dungeons_woth_limit'] and \
+        location.parent_region.dungeon) and \
+        not (location.name in world.hint_type_overrides['woth']) and \
```
Use `a not in b` notation throughout.

mracsys replied:
Done

Zannick commented on World.py:
```diff
@@ -108,7 +110,61 @@ def __init__(self, id, settings):
+                if not all(sub_key in sub_keys for sub_key in self.hint_dist_user['distribution'][key]):
+                    hint_dist_valid = False
+        if not hint_dist_valid:
+            raise InvalidFileException("Hint distributions require all hint types be present in the distro (trial, always, woth, barren, item, song, overworld, dungeon, entrance, sometimes, random, junk, named-item). If a hint type should not be shuffled, set its order to 0. Hint type format is \"type\": { \"order\": 0, \"weight\": 0.0, \"fixed\": 0, \"stones\": 0 }")
```
Split long strings into multiple lines, please.

mracsys replied:
Done

Zannick commented on World.py:
```diff
@@ -108,7 +110,61 @@ def __init__(self, id, settings):
+        self.hint_text_overrides = {}
+        for loc in self.hint_dist_user['add_locations']:
+            if 'text' in loc:
+                # Arbitrarily throw an error at 80 characters to prevent overfilling the text box.
```
There's a thought. Is there actually a specific limit?

mracsys replied:
I'm not sure. Max length in my new hints is 71 characters for `Spirit Temple MQ Map Room Enemy Chest`. I tested it in-game with a Triforce Piece in it, and there was plenty of room left visually. 80 seemed like a decent cutoff.

Zannick commented on data/Hints/useless.json:
```diff
@@ -0,0 +1,27 @@
+    "remove_items":          [],
+    "dungeons_woth_limit":   2,
+    "dungeons_barren_limit": 1,
+    "named_items_required":  "False",
```
Going by the line `if len(world.item_hints) > 0 and world.hint_dist_user['named_items_required']:` you want `false` here, not `"False"`. The latter is a non-empty string which will evaluate to `True`. The other files should likewise have `true` here to avoid confusion.

mracsys replied:
Changed to booleans as suggested, tested on useless, tournament, and custom distributions to make sure all cases work as expected.

Test plando:
```
{
    "settings": {
      "item_hints": [
            "Progressive Strength Upgrade",
            "Progressive Strength Upgrade",
            "Lens of Truth"
        ]
    },
    "randomized_settings": {},
    "starting_items":      {},
    "item_pool":           {},
    "entrances":           {},
    "locations":           {}
}
```

mracsys commented:
Most of them are not enabled unless a custom hint distribution is active (in the "exclude" hint group instead of "sometimes" or its sub-types). I stayed away from the clearer hints for current "enabled" locations outside of bugfixes for the DMC skull and scrub.

For unique text, I tried to make each as unique as possible with the shear amount of them. Some of them are duplicated where they're in the same room with the same requirements (e.g. Deku Tree Compass room). Others, like the scrub grottos, don't have any cues for which is which in ER. I wasn't consistent with this though, as the generic grottos got some entrance-specific text. All of these are meant as placeholders to allow the plando changes to work with all locations. I'm open to any and all suggestions.

mracsys commented on HintList.py:
```diff
@@ -198,79 +221,763 @@ def getRequiredHints(world):
+
+    'Deku Tree MQ Map Chest':                                      ("in the #center of the Deku Tree# lies", None, 'exclude'),
+    'Deku Tree MQ Compass Chest':                                  ("in #treasure guarded by a large spider# in the Deku Tree is", None, 'exclude'),
+    'Deku Tree MQ Slingshot Chest':                                ("#pillars of wood# in the Deku Tree leads to", None, 'exclude'),
```
Done

mracsys commented:
A discussion came up on the discord around the Skull Mask hint. I changed it to an always hint for the tournament distribution regardless of open/closed Kakariko Gate. I've reverted it in the last commit.

Zannick commented:
r0b, do you want to discuss all the new hints here, or would it be better to address them in a separate issue/PR? The new ones pretty much all have to be explicitly enabled by the hintdist plando to be used.

r0bd0g commented:
I kind of wanted to just go through and nitpick every single text lol but there's so much of it. I started but mostly ended up only complaining about texts that weren't even changed in this PR lol. This is as far as I got:

KF Links House Cow - It says "Malon's obstacle course leads to" the item, but what Malon's obstacle course leads to is a cow. So I don't like that the fact that the hint implies the item comes from Malon and not from the cow. This doesn't lead to any kind of confusion about what check is being referred to, it's just that the text is not accurate when written this way. Something like "A cow in [filename]'s house gifts" would be better, even though that doesn't mention the obstacle course at all, since that makes it clear the prize comes from a cow.

Sheik at Colossus - This hint text was never all that great, and with the possibility of ER it doesn't really make sense anymore.

Fire Temple Scarecrow Chest - I don't like the way "Pierre atop the Fire Temple hides" reads. It feels like there are commas missing. Like maybe it should be "Pierre, atop the Fire Temple, hides". But that would read better as "atop the Fire Temple, Pierre hides". But then you have to wonder why it was changed from "High in the Fire Temple, Pierre hid" at all?

Water Temple MQ GS Triple Wall Torch - The text mentions a skulltula behind a gate. The equivalent location in vanilla makes sense for this text, and they probably should share the same text, but I worry b/c MQ Water has another skulltula behind a gate as well (Lizalfos hallway). Maybe if you just add the word "deep" in there somewhere it would help to make it clear it's a basement skulltula at least.


You added a number of non-clearer hints that in general I don't think are necessary, though I do like a couple of them.

I'm gonna call out a few though that I think are too unclear: DMT Biggoron, Kak Anju as Child (bundling birds, specifically), Graveyard Heart Piece Grave Chest, GC Maze Left Chest.

Gerudo Training Grounds [MQ] Underwater Silver Rupee Chest - I mean if you want to add a new hint text here, that's fine, but I don't get the impression that either of these hints is more or less clear than the other, so I would say just to combine them together.


And finally let's go through all the new hint texts.

Deku Theater Mask of Truth & Deku Theater Skull Mask - I dunno, there's something about the phrase "exposing scrubs" that I don't like the sound of... Regardless, the "E" at the start of that hint text should be lowercase.

KF Kokiri Sword Chest - I don't understand "a child's toy chest holds". And for "inside a butter cutter box lies", I don't think hint texts should mention the vanilla contents in this way. The clear hint you have here is great though should suffice regardless of the clearer hints setting. I'd suggest maybe going with "hidden treasure of the Kokiri" as that is what the pick-up text refers to the Kokiri Sword as.

Graveyard Hookshot Chest - The non-clearer text here I think is too unclear.

mracsys commented:
Thanks for your input, r0b! I'm not hard set on most of these. Who is the target audience for the non-clear hints, and how accessible should they be? The ones I wrote can get obscure, but there is always some connection back to OoT lore, acquisition requirements, and/or the surrounding environment.

I'm treating all of the "hidden" hints as first drafts pending community review. They are easy to override in the hint distro in the short term if any are chosen to be included in races.

r0bd0g commented:
The location texts were generally not super cryptic, I dunno how to explain. It's not that we ever really tried to actively confuse anybody, just inject a little bit or lore or wordplay. Uh... instead we just confused everybody lol. Quite a few of the non-clearer hints for item names do go full cryptic. It's hard sometimes to think up alternative names for all those items and have no confusion... For locations though, a lot of the clearer hints are used for things like "high mountain"/"Fire Temple", turning the dungeon flavour text into the dungeon name. A lot of them are just more direct, turning "gambling" into "Treasure Chest Game". I dunno if many of them go "full cryptic" if that makes any sense. At this point nobody uses non-clearer anymore anyway. Making it more cryptic I don't think is gonna help things...

mracsys commented:
For the hookshot chest, maybe "a chest guarded by a #speedy spectre# holds"?

I removed the other questionable new hints. Personally, I think the existing hint changes should go in a new issue.

r0bd0g commented:
Seems good. I think at least before any release I'd like to go through all of it, maybe make a bit pastebin of nitpicks.
Gotta be careful not to commit to too much though, lol, still have the MQ logic update to get through.

Zannick commented:
Maybe we should rethink what hints are; instead of having different levels of clarity per hint, we have hints with different levels of precision. The clear hints are supposed to be "this location has this item". We could have less precise hints that say "this area has this item".

Just a thought for now; I'll be merging this PR in today.
